### PR TITLE
Dev 1.5.0 462 Separate TensorFlow 1.x and TensorFlow 2.x tests

### LINKED ISSUE #462: Treat Tensorflow1 and Tensorflow2 as two separate framework in tests
**Is your feature request related to a problem? Please describe.**
Currently TS1 and TS2 are considered subsets of the tensorflow framework. Because of this, many tests currently have if statements checking whether one version of tensorflow is used or another. This breaks the notion of framework agnostic tests and adds unecessary lines of test code.

**Describe the solution you'd like**
We can simply consider TS1 and TS2 as two separate frameworks and instead take advantage of the existing pytest framework fixture and all available markers to control the behavior of the tests with respect to the tensorflow versions used.

### PR DESCRIPTION
**Pull request by killianlevacher**
# Description
This PR addresses issue #462 which main purpose involves removing all duplicated boilerplate within the tests dealing with identifying which version of Tensorflow is currently being ran and treating TF1 and TF2 as two separate frameworks. This means we can avail of all the pytest functionality we already created for managing frameworks

Fixes #462 

## Type of change

Please check all relevant options.

- [X] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [ ] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [ ] Test A
- [ ] Test B

**Test Configuration**:
- OS
- Python version
- ART version or commit number
- TensorFlow / Keras / PyTorch / MXNet version

# Checklist

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: conftest.py [hunk 1]
```diff
@@ -24,6 +24,7 @@
 import numpy as np
 import pytest
 import requests
+import tensorflow as tf
```
Comment (beat-buesser) [inline conftest.py:27]: Please move import closer to application, the top-level import would create dependency.

File: run_tests.sh [hunk 2]
```diff
@@ -6,47 +6,48 @@ export TF_CPP_MIN_LOG_LEVEL="3"
...
+  if [[ $? -ne 0 ]]; then exit_code=1; echo "Failed attacks/evasion/test_shadow_attack.py"; fi
 
-#NOTE: All the tests should be ran within this loop. All other tests are legacy tests that must be
-# made framework independent to be incorporated within this loop
-mlFrameworkList=("tensorflow" "keras" "pytorch" "scikitlearn" "mxnet" "kerastf")
-for mlFramework in "${mlFrameworkList[@]}"; do
-  echo "Running tests with framework $mlFramework"
+  pytest --cov-report=xml --cov=art --cov-append  -q -vv tests/attacks/inference/test_model_inversion.py --mlFramework=$mlFramework --skip_travis=True --durations=0
+  pytest --cov-report=xml --cov=art --cov-append  -q -vv tests/attacks/inference/test_attribute_inference.py --mlFramework=$mlFramework --skip_travis=True --durations=0
+  if [[ $? -ne 0 ]]; then exit_code=1; echo "Failed attacks/inference tests"; fi
```
Comment (beat-buesser) [inline run_tests.sh:39]:
```suggestion
  pytest --cov-report=xml --cov=art --cov-append  -q -vv tests/attacks/inference/test_model_inversion.py --mlFramework=$mlFramework --skip_travis=True --durations=0
  if [[ $? -ne 0 ]]; then exit_code=1; echo "Failed attacks/inference/model_inversion tests"; fi
  pytest --cov-report=xml --cov=art --cov-append  -q -vv tests/attacks/inference/test_attribute_inference.py --mlFramework=$mlFramework --skip_travis=True --durations=0
  if [[ $? -ne 0 ]]; then exit_code=1; echo "Failed attacks/inference/attribute_inference tests"; fi
```

File: tests/attacks/evasion/test_shadow_attack.py [hunk 3]
```diff
@@ -58,7 +58,7 @@ def test_generate(art_warning, fix_get_mnist_subset, image_dl_estimator_for_atta
 
         x_train_mnist_adv = attack.generate(x=x_train_mnist[0:1], y=y_train_mnist[0:1])
 
-        assert np.max(np.abs(x_train_mnist_adv - x_train_mnist[0:1])) == pytest.approx(0.34966960549354553, 0.06)
+        assert np.max(np.abs(x_train_mnist_adv - x_train_mnist[0:1])) == pytest.approx(0.38116083, 0.06)
```
Comment (beat-buesser) [inline tests/attacks/evasion/test_shadow_attack.py:61]: Why did this expectation value change?

File: tests/defences/preprocessor/test_spatial_smoothing.py [hunk 4]
```diff
@@ -29,36 +29,36 @@
...
+#     test_output = test_input.copy()
+#     return test_input, test_output
+#
+#
+# @pytest.fixture
+# def tabular_batch():
+#     """
+#     Create tabular data fixture of shape (batch_size, features).
+#     """
+#     return np.zeros((2, 4))
```
Comment (beat-buesser) [inline tests/defences/preprocessor/test_spatial_smoothing.py:61]: Can we remove these commented lines since they are now in `conftest.py`?

File: tests/defences/preprocessor/test_spatial_smoothing_pytorch.py [hunk 5]
```diff
@@ -29,37 +29,37 @@
...
+#     test_output = test_input.copy()
+#     return test_input, test_output
+#
+#
+# @pytest.fixture
+# def tabular_batch():
+#     """
+#     Create tabular data fixture of shape (batch_size, features).
+#     """
+#     return np.zeros((2, 4))
```
Comment (beat-buesser) [inline tests/defences/preprocessor/test_spatial_smoothing_pytorch.py:62]: Can we remove these commented lines since they are now in `conftest.py`?

File: tests/defences/preprocessor/test_spatial_smoothing_tensorflow.py [hunk 6]
```diff
@@ -29,63 +29,61 @@
...
+#     test_output = test_input.copy()
+#     return test_input, test_output
+#
+#
+# @pytest.fixture
+# def tabular_batch():
+#     """
+#     Create tabular data fixture of shape (batch_size, features).
+#     """
+#     return np.zeros((2, 4))
```
Comment (beat-buesser) [inline tests/defences/preprocessor/test_spatial_smoothing_tensorflow.py:62]: Can we remove these commented lines since they are now in `conftest.py`?

File: run_tests.sh [hunk 7]
```diff
@@ -6,47 +6,48 @@ export TF_CPP_MIN_LOG_LEVEL="3"
...
 for mlFramework in "${mlFrameworkList[@]}"; do
-  echo "Running tests with framework $mlFramework"
-  pytest -q -vv tests/attacks/inference/ --mlFramework=$mlFramework --skip_travis=True --durations=0
+  echo "#######################################################################"
+  echo "############## Running tests with framework $mlFramework ##############"
+  echo "#######################################################################"
+
+  #TODO FIX Failing with Keras at test_membership_inference.test_black_box_keras_loss
+  pytest --cov-report=xml --cov=art --cov-append -q -vv tests/attacks/inference/test_membership_inference.py --mlFramework=$mlFramework --skip_travis=True --durations=0
   if [[ $? -ne 0 ]]; then exit_code=1; echo "Failed attacks/inference tests"; fi
```
Comment (beat-buesser) [inline run_tests.sh:17]: This TODO has been fixed in the meantime with PR #651 on `dev_1.4.2` and `dev_1.5.0`. Can we now move this run inside of the general loop?

**Review by beat-buesser — CHANGES_REQUESTED** [review]
Hi @killianlevacher , thank you very much for this pull request. It facilitates handling the different versions of TensorFlow. I have a few comments and requests. Could you please take a look?

File: run_tests.sh (hunk 7 above)
Comment (killianlevacher) [reply run_tests.sh:17]: ok added it to the main loop

File: tests/attacks/evasion/test_shadow_attack.py (hunk 3 above)
Comment (killianlevacher) [reply tests/attacks/evasion/test_shadow_attack.py:61]: oh ye that was a nasty one ! :D so while I was trying to bring all the tests that were not in the main loop, this test kept failing and I couldn't understand why because I hadn't changed anything at all in there. Then I discovered that that test was never run because it had the `@pytest.mark.only_with_platform("pytorch")` marker but was in a loop which never ran the tests with pytorch which is why it was not failing before and failing for pytorch in the main loop. So since I haven't made any changes to the models in this PR I it was ok if I updated the value that the pytorch model actually outputs to make sure that test can now be run as part of the main loop. I'm not sure if this makes sense?

**Review by beat-buesser — APPROVED** [review]
